### DISCUSSION TITLE
Create More Flexable SessionMiddleware

### DISCUSSION THREAD

**Discussion by lannuttia** [encode/starlette#2256] [Ideas]
Hey all, I am currently working on an application that I am planning on phasing out as part of my day job that utilizes JWTs. In order to facilitate interoperability between the legacy and new application, it would be really nice to have a JWT session middleware built into the starlette upstream itself. I'll be honest, I was a little surprised that this wasn't available in the upstream already and was relegated to third-party packages. I realize that there are third-party packages that provide similar JWT session middleware such as starlette-jwt and starlette-authlib. However, I'm wary of utilizing tools like those out of fear of them being abandoned. I had already created a #2255 with support for cookie based JWT sessions and that was not merged. I recognize that my original implementation was limited as presented in that PR but after looking into the existing session middleware and creating the mostly similar JWT session middleware in #2255, I think it would be doable to create a flexible session middleware that could support cookie-based and authorization header based sessions with support for JWT and the existing signed token session middleware.  From what I've seen, all these session middleware (in starlette itself as well as in the third-party packages) all follow a very similar structure.

I would envision this flexible session middleware being largely similar to the existing session middleware but with something like an encode/decode backend for handling encoding and decoding generically and a storage middleware that is responsible for storing the resulting token after encode and retrieving the token for decoding. The encode/decode backend I would envision just needing to be a class that implements encode and decode methods. As for the storage backend, I would envision it just needing to be a class that implements a store and retrieve method.

**Comment by iudeen** [comment]
I saw your PR. It can very well be an independent package, that can be used by not just starlette, but other ASGI servers. 

I'm not sure why you think third-party packages may get abandoned. If it's a feature that's popular, and if a lot of people would use it, it will be maintained by the community. I understand you may not have the time to maintain it, but creating a package is a starter.

**Reply by lannuttia** [reply]
I have created a package that I think will allow for me to accomplish what I had specified in this discussion. It is currently only an alpha release because I still haven't decided whether I want the backend interfaces to keep the API that I have given them right now but I think what I've got is close to where I want it. Is this something that I could get merged into starlette once the API is finalized? If this is still not something that you all are interested in getting distributed with starlette itself, is this something that could be referenced in the third-party middleware documentation once the API has been finalized?

[The project](https://github.com/lannuttia/starlette-session-middleware)
